hack/ci-e2e.sh

```diff
@@ -176,17 +176,17 @@ for overlay in "${BMO_OVERLAYS[@]}"; do
 done
 
 for overlay in "${IRONIC_OVERLAYS[@]}"; do
-  echo "IRONIC_HTPASSWD=$(htpasswd -n -b -B "${IRONIC_USERNAME}" "${IRONIC_PASSWORD}")" > \
-    "${overlay}/ironic-htpasswd"
+  echo "${IRONIC_USERNAME}" > "${overlay}/ironic-username"
+  echo "${IRONIC_PASSWORD}" > "${overlay}/ironic-password"
```

Comment:
I'm confused by this. We used to put an htpasswd here, which contained the username in plain text and the password hashed by bcrypt. That means the password is not directly readable but still verifiable. With this change we have both username and password as plain text which is definitely not best practice.
Is it really the intention to mount them as plain text? Why?

Reply:
I was reading about kustomize secretGeneration using files [here](https://kubernetes.io/docs/tasks/configmap-secret/managing-secret-using-kustomize/#:~:text=In%20all%20cases%2C%20you%20don%27t%20need%20to%20base64%20encode%20the%20values.) Mounting the password as plain text does seem wrong, though. Will ironic-image be able to generate an htpasswd with the ironic-password if its given a hash of the password? It could, but it wouldn't be correct, but more secure overall? What do you think?

Reply:
Clarified offline, commenting for closure.
We should essentially revert https://github.com/metal3-io/ironic-image/pull/482/files so that ironic does not generate the htpasswd. Instead, we put that directly in the secret and mount it as a file that ironic can use.

Reply:
Sorry, but absolutely no. Mounting secrets is the normal and documented way to provide information into pods. The pod is responsible for any fancy format it uses. Passing htpasswd in any form is just increasing the load on users trying to install ironic (and provides for some terrible code in ironic-standalone-operator), especially since ironic does not accept any htpasswd (only certain ciphers).

Reply:
Oh, and another consideration. We're working on liveness/readiness checks that do more than just make sure the API exists. These will require credentials. I left a TODO until Mahnoor's work is done: https://github.com/metal3-io/ironic-image/blob/main/scripts/ironic-probe.j2#L7-L9

Reply:
On the bright side, when we split away configuration as you're planning @lentzi90, the secret will only be mounted for the short duration of the init container.

Reply:
Fair enough! I understand the reasoning. It was just hard to understand from the initial description that hinted at *better* security. Going from environment variable -> mounted file I can see as some kind of improvement to security, but to then at the same time get rid of bcrypt raised red flags for me.

I'm fine storing it as plain text. However, I must then insist that we rename the secret. It cannot be named `ironic-htpasswd` "because ironic will make an htpasswd out of it". It will be too confusing.

Reply:
Yeah, no disagreement on the naming part.

Reply:
Hey, sorry to reopen this discussion, but I think there is potentially a better way of doing this that still meets all the future requirements (for clarity, I mean do these steps instead of any changes that are in progress or have merged):

- Leave the startironic logic as it was for now
- Mount the htpasswd from the secret in the same location that the startironic script would have saved it instead of passing it as an env var. (This solves the problem without breaking backward compatibility for any existing deployments that pass it as an env var.)
- In future, add a separate init container setup script that takes the plaintext password (mounted from a secret) and generates an htpasswd file saved in the correct location for startironic to read it
- Further in the future, generate a second username/password in the init container for use with by the readiness check, and add that into the same htpasswd file. Save _this_ password as plaintext to pass through to the ironic container for the use of the readiness check. (There's no authz in ironic, but this is marginally more secure as it gets regenerated every time the pod starts.)

I think that would avoid breaking backward compat, while still allowing us to make things easier to deploy over time (even without ironic-operator), and not blocking us from adding readiness checks either.

Reply:
Sounds reasonable to me! Do you mean that we should revert https://github.com/metal3-io/ironic-image/pull/482 then?

Reply:
Sorry, s/startironic/[runironic](https://github.com/metal3-io/ironic-image/blob/main/scripts/runironic)/. i.e. the ironic-image container when used to run ironic (as opposed to when it is used to run httpd, or ironic-inspector, or dnsmasq).

> - We all agree to mount the htpasswd under /auth/ironic/htpasswd in ironic-image (ironic-image now no longer accepts the IRONIC_HTPASSWD environment variable and reads from the file instead).

I was thinking both.

- If the user passes nothing, then `/auth/ironic/htpasswd` doesn't exist
- If the user passes IRONIC_HTPASSWD then it gets written to `/auth/ironic/htpasswd`
- If the user doesn't pass IRONIC_HTPASSWD but mounts a file at `/auth/ironic/htpasswd` then it's already there

Reply:
So we just provide a new way in ironic-image for htpasswd to be provided (via file), and security-conscious consumers can use the new way. Does that fully solve the inherent complaint about ironic-image being insecure? Maybe it addresses it sufficiently?

Reply:
I don't think we can make the change without a compatibility layer still. The htpasswd is put into `/etc/ironic/htpasswd` and `/etc/ironic/htpasswd-rpc` conditionally and some options are set. What you suggest @zaneb makes certain sense, but it requires reworking runironic either way.

@MahnoorAsghar we probably don't want to have 2 ways to pass htpasswd long-term. IRONIC_HTPASSWD will need to be kept for backward compatibility for some time.

Reply:
So for now, we can keep both options, and eventually phase out the environment variable IRONIC_HTPASSWD.

Reply:
> Does that fully solve the inherent complaint about ironic-image being insecure? Maybe it addresses it sufficiently?

I think the complaint is about how we expose data from a Secret (i.e. as an environment variable), not how ironic-image receives it. So as long as we switch to the new method (i.e. mounting as file), I think the complaints go away.

Reply:
> it requires reworking runironic either way

Yes, that's fine, TBH I haven't looked into any of the details. The mount location could even be somewhere different (as long as it is clearly documented) and we copy the file to where it is needed at startup if it exists, or generate the file where it is needed from the env vars if they exist. Whatever implementation makes sense.